Updates preload scanner guide (#8846)
GoogleChrome/web.dev: src/site/content/en/blog/preload-scanner/index.md

With the `rel=preload` hint, the LCP candidate is discovered sooner, lowering the LCP time. While that hint helps fix this issue, the better option may be to assess whether or not your image LCP candidate _has_ to be loaded from CSS. With an `<img>` tag, you'll have more control over loading an image that's appropriate for the viewport while allowing the preload scanner to discover it.

## Inlining too much stuff

> Inlining is a practice that places a resource inside of the HTML. You can inline stylesheets in `<style>` elements, scripts in `<script>` elements, and virtually any other resource using [base64 encoding](https://developer.mozilla.org/docs/Glossary/Base64).

Reviewer: ```suggestion
Inlining is a practice that places a resource inside of the HTML. You can inline stylesheets in `<style>` elements, scripts in `<script>` elements, and virtually any other resource using [Base64 encoding](https://developer.mozilla.org/docs/Glossary/Base64).
```

Reviewer: Ignore. Style guide is lowercase b for base64

> Inlining resources is faster than downloading them because a separate request isn't issued for the resource. It's right in the document, and loads instantly. However, there are drawbacks:

Reviewer: ```suggestion
Inlining resources is often faster than downloading them because a separate request isn't issued for the resource. It's right in the document, and loads instantly. However, there are drawbacks:
```

> - If you're not caching your HTML&mdash;and you just can't if the HTML response is dynamic&mdash;the inlined resources are never cached. This affects performance because the inlined resources aren't reusable.

Reviewer: Even if you ARE caching your HTML, if you inline your styles instead of including at styles.css, then another page on the site doesn't benefit of that caching.

Worth adding a third point on that?

Take [this page](https://preload-scanner-fights.glitch.me/inline-nothing.html) as an example. In certain conditions the LCP candidate is the image at the top of the page, and the CSS is in a separate file loaded by a `<link>` element. The page also uses four web fonts which are requested as separate files from the CSS resource.

<figure>
  {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/GY0b2sh3wRt0CVpwdPvr.png", alt="A WebPageTest network waterfall chart of page with an external CSS file with four fonts referenced in it. The LCP candidate image is discovered by the preload scanner in due course.", width="800", height="297", loading="lazy" %}
  <figcaption>
    <strong>Fig. 12:</strong> A WebPageTest network waterfall chart of <a href="https://preload-scanner-fights.glitch.me/inline-nothing.html" rel="noopener">a web page</a> run on Chrome on a mobile device over a simulated 3G connection. The page's LCP candidate is an image loaded from an <code>&lt;img&gt;</code> element, but is discovered by the preload scanner because the CSS and the fonts required for the page load in separate resources, which avoids preventing the preload scanner from doing its job.
  </figcaption>
</figure>

> Now what happens if we inline the CSS _and_ inline all the fonts as base64 resources?

Reviewer: ```suggestion
Now what happens if we inline the CSS _and_ inline all the fonts as Base64 resources?
```

<figure>
  {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/ommDo1DKQUu1GxJ3YPvq.png", alt="A WebPageTest network waterfall chart of page with an external CSS file with four fonts referenced in it. The preload scanner is delayed significantly from discovering the LCP image .", width="800", height="297", loading="lazy" %}
  <figcaption>
    <strong>Fig. 13:</strong> A WebPageTest network waterfall chart of <a href="https://preload-scanner-fights.glitch.me/inline-everything.html" rel="noopener">a web page</a> run on Chrome on a mobile device over a simulated 3G connection. The page's LCP candidate is an image loaded from an <code>&lt;img&gt;</code> element, but the inlining of the CSS and its four font resources delays the preload scanner from discovering the image until those resources are fully parsed.
  </figcaption>
</figure>

The impact of inlining yields negative consequences for LCP in this example&mdash;and for performance in general. The version of the page that doesn't inling anything paints the LCP image in about 3.5 seconds. The page that inlines everything doesn't paint the LCP image until roughly 6.4 seconds.

> To be candid, there's more at play here than just the preload scanner. Inlining fonts is not a great strategy in part because external font resources aren't downloaded unless they're determined necessary by the CSSOM. When those fonts are inlined as base64, however, they're downloaded whether they're needed or not.

Reviewer: ```suggestion
To be candid, there's more at play here than just the preload scanner. Inlining fonts is not a great strategy because Base64 is an inefficient format for binary resources. Additionally, because external font resources aren't downloaded unless they're determined necessary by the CSSOM but when those fonts are inlined as Base64, they're downloaded whether they're needed or not.
```

Could a preload improve things here? Sure. You _could_ preload the LCP image and reduce LCP time, but bloating your potentially uncacheable HTML with inlined resources has other negative performance consequences. [First Contentful Paint (FCP)](/fcp/) is also affected by this pattern. In the version of the page where nothing is inlined, FCP is roughly 2.7 seconds. In the version where everything is inlined, FCP is roughly 5.1 seconds.

> Be very careful with inlining stuff into HTML, especially base64-encoded resources. Don't do it if you can help it. If you can't help it, inline as little as possible. Inlining too much is playing with fire.

Reviewer: ```suggestion
Be very careful with inlining stuff into HTML, especially Base64-encoded resources. In general it is not recommended, except for very small resources. So, inline as little as possible&mdash;inlining too much is playing with fire.
```

With the `rel=preload` hint, the LCP candidate is discovered sooner, lowering the LCP time. While that hint helps fix this issue, the better option may be to assess whether or not your image LCP candidate _has_ to be loaded from CSS. With an `<img>` tag, you'll have more control over loading an image that's appropriate for the viewport while allowing the preload scanner to discover it.

> ## Inlining too much stuff

Reviewer: "Stuff" feels a little awkward here.

```suggestion
## Inlining too many resources
```

Inlining resources is often faster than downloading them because a separate request isn't issued for the resource. It's right in the document, and loads instantly. However, there are drawbacks:

- If you're not caching your HTML&mdash;and you just can't if the HTML response is dynamic&mdash;the inlined resources are never cached. This affects performance because the inlined resources aren't reusable.
- If you inline too much, you delay the preload scanner from discovering resources in the document, because parsers for CSS and scripts have to do their thing before the preload scanner can continue on to discover other resources in the document.
- Even if you can cache the HTML, inlined resources aren't shared between documents. This reduces caching efficiency compared to external files that can be cached and reused across an entire origin.

Take [this page](https://preload-scanner-fights.glitch.me/inline-nothing.html) as an example. In certain conditions the LCP candidate is the image at the top of the page, and the CSS is in a separate file loaded by a `<link>` element. The page also uses four web fonts which are requested as separate files from the CSS resource.

<figure>

>   {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/GY0b2sh3wRt0CVpwdPvr.png", alt="A WebPageTest network waterfall chart of page with an external CSS file with four fonts referenced in it. The LCP candidate image is discovered by the preload scanner in due course.", width="800", height="297", loading="lazy" %}

Author: Yep, noted. Going to throw in replacement images, as I have to rerun tests (argh).

>   {% Img src="image/jL3OLOhcWUQDnR4XjewLBx4e3PC3/GY0b2sh3wRt0CVpwdPvr.png", alt="A WebPageTest network waterfall chart of page with an external CSS file with four fonts referenced in it. The LCP candidate image is discovered by the preload scanner in due course.", width="800", height="297", loading="lazy" %}

Author: This is now fixed!
